tests/components/statistics/test_sensor.py

```diff
@@ -1701,3 +1702,62 @@ async def test_device_id(
     statistics_entity = entity_registry.async_get("sensor.statistics")
     assert statistics_entity is not None
     assert statistics_entity.device_id == source_entity.device_id
+
+
+async def test_update_before_load(recorder_mock: Recorder, hass: HomeAssistant) -> None:
+    """Verify that updates happening before reloading from the database are handled correctly."""
+
+    current_time = dt_util.utcnow()
+
+    # enable and pre-fill the recorder
+    await hass.async_block_till_done()
+    await async_wait_recording_done(hass)
+
+    with (
+        freeze_time(current_time) as freezer,
+    ):
+        for value in VALUES_NUMERIC_LINEAR:
+            hass.states.async_set(
+                "sensor.test_monitored",
+                str(value),
+                {ATTR_UNIT_OF_MEASUREMENT: UnitOfTemperature.CELSIUS},
+            )
+            await hass.async_block_till_done()
+            current_time += timedelta(seconds=1)
+            freezer.move_to(current_time)
+
+        await async_wait_recording_done(hass)
+        # create the statistics component, get filled from database
+        assert await async_setup_component(
+            hass,
+            "sensor",
+            {
+                "sensor": [
+                    {
+                        "platform": "statistics",
+                        "name": "test",
+                        "entity_id": "sensor.test_monitored",
+                        "state_characteristic": "average_step",
+                        "max_age": {"seconds": 10},
+                    },
+                ]
+            },
+        )
+        # this value is probably going to be ignored, since loading from the database has
+        # most likely hasn't finished yet
+        # if this value would be added before loading the historic data
+        # it would mess up the order of the internal queue which is supposed to be sorted by time
+        hass.states.async_set(
+            "sensor.test_monitored",
+            "10",
+            {ATTR_UNIT_OF_MEASUREMENT: DEGREE},
+        )
+        await hass.async_block_till_done()
+
+    # depending on timing we will either end up with a buffer of [1 .. 9] or [1 .. 10]
+    # what may not happen is that the 10 will be added somewhere in between
+    # so we compute average_step for either 1 .. 9 or 1 .. 10
+    # this leads to 1+2+3+4+5+6+7+8/8 = 4.5
+    #            or 1+2+3+4+5+6+7+8+9/9 = 5
+    avg: float = float(hass.states.get("sensor.test").state)
+    assert avg == pytest.approx(4.5) or avg == pytest.approx(5.0)
```

Comment:
I don't like this, our tests should be deterministic. We should instrument the test as needed so we can reliably trigger and test both cases:
- A state change happens while we're reading from the database
  - We should be able simulate this case with the change below
- A state change does not happen while we're reading from the database
  - Maybe we don't need a special test for this, this is the normal case, right?

```diff
diff --git a/tests/components/statistics/test_sensor.py b/tests/components/statistics/test_sensor.py
index 43730fcc665..29a102b1112 100644
--- a/tests/components/statistics/test_sensor.py
+++ b/tests/components/statistics/test_sensor.py
@@ -2,9 +2,11 @@

 from __future__ import annotations

+from asyncio import Event as AsyncioEvent
 from collections.abc import Sequence
 from datetime import datetime, timedelta
 import statistics
+from threading import Event
 from typing import Any
 from unittest.mock import patch

@@ -12,7 +14,7 @@ from freezegun import freeze_time
 import pytest

 from homeassistant import config as hass_config
-from homeassistant.components.recorder import Recorder
+from homeassistant.components.recorder import Recorder, history
 from homeassistant.components.sensor import (
     ATTR_STATE_CLASS,
     SensorDeviceClass,
@@ -44,6 +46,7 @@ from homeassistant.core import HomeAssistant
 from homeassistant.helpers import device_registry as dr, entity_registry as er
 from homeassistant.setup import async_setup_component
 from homeassistant.util import dt as dt_util
+from homeassistant.util.async_ import run_callback_threadsafe

 from tests.common import MockConfigEntry, async_fire_time_changed, get_fixture_path
 from tests.components.recorder.common import async_wait_recording_done
@@ -1728,30 +1731,43 @@ async def test_update_before_load(recorder_mock: Recorder, hass: HomeAssistant)

         await async_wait_recording_done(hass)
         # create the statistics component, get filled from database
-        assert await async_setup_component(
-            hass,
-            "sensor",
-            {
-                "sensor": [
-                    {
-                        "platform": "statistics",
-                        "name": "test",
-                        "entity_id": "sensor.test_monitored",
-                        "state_characteristic": "average_step",
-                        "max_age": {"seconds": 10},
-                    },
-                ]
-            },
-        )
-        # this value is probably going to be ignored, since loading from the database has
-        # most likely hasn't finished yet
-        # if this value would be added before loading the historic data
-        # it would mess up the order of the internal queue which is supposed to be sorted by time
-        hass.states.async_set(
-            "sensor.test_monitored",
-            "10",
-            {ATTR_UNIT_OF_MEASUREMENT: DEGREE},
-        )
+        state_changes_during_period_called_evt = AsyncioEvent()
+        state_changes_during_period_stall_evt = Event()
+
+        real_state_changes_during_period = history.state_changes_during_period
+        def mock_state_changes_during_period(*args, **kwargs):
+            states =  real_state_changes_during_period(*args, **kwargs)
+            hass.loop.call_soon_threadsafe(state_changes_during_period_called_evt.set)
+            state_changes_during_period_stall_evt.wait()
+            return states
+
+        with patch("homeassistant.components.statistics.sensor.history.state_changes_during_period", mock_state_changes_during_period):
+            assert await async_setup_component(
+                hass,
+                "sensor",
+                {
+                    "sensor": [
+                        {
+                            "platform": "statistics",
+                            "name": "test",
+                            "entity_id": "sensor.test_monitored",
+                            "state_characteristic": "average_step",
+                            "max_age": {"seconds": 10},
+                        },
+                    ]
+                },
+            )
+            # this value is probably going to be ignored, since loading from the database has
+            # most likely hasn't finished yet
+            # if this value would be added before loading the historic data
+            # it would mess up the order of the internal queue which is supposed to be sorted by time
+            await state_changes_during_period_called_evt.wait()
+            hass.states.async_set(
+                "sensor.test_monitored",
+                "10",
+                {ATTR_UNIT_OF_MEASUREMENT: DEGREE},
+            )
+            state_changes_during_period_stall_evt.set()
         await hass.async_block_till_done()

     # depending on timing we will either end up with a buffer of [1 .. 9] or [1 .. 10]
@@ -1760,4 +1776,4 @@ async def test_update_before_load(recorder_mock: Recorder, hass: HomeAssistant)
     # this leads to 1+2+3+4+5+6+7+8/8 = 4.5
     #            or 1+2+3+4+5+6+7+8+9/9 = 5
     avg: float = float(hass.states.get("sensor.test").state)
-    assert avg == pytest.approx(4.5) or avg == pytest.approx(5.0)
+    assert avg == pytest.approx(4.5)
```

Tested with `pytest -x --count=500 tests/components/statistics/test_sensor.py::test_update_before_load`